heartbeat/scheduler/scheduler.go

```diff
@@ -138,282 +178,102 @@ var ErrAlreadyStopped = errors.New("attempted to add job to already stopped sche
-			debugf("done")
-			return
-
-		}
+		schedNextRun()
 	}
-}
 
-func reschedActive(j *job, now time.Time) {
-	logp.Info("Scheduled job '%v' already active.", j.id)
-	if !now.Before(j.next) {
-		j.next = j.schedule.Next(j.next)
+	if sched.RunOnInit() {
+		go taskFn(time.Now())
+	} else {
+		schedNextRun()
 	}
-}
 
-func (s *Scheduler) startJob(j *job) {
-	j.running++
-	j.next = j.schedule.Next(j.next)
-	debugf("Start job '%v' at %v.", j.id, time.Now())
+	return jobCtxCancel, nil
+}
 
-	s.runTask(task{j, j.fn})
+// runRecursiveJob runs the entry point for a job, blocking until all subtasks are completed.
+// Subtasks are run in separate goroutines.
+func (s *Scheduler) runRecursiveJob(jobCtx context.Context, task TaskFunc) {
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	s.runRecursiveTask(jobCtx, task, wg)
+	wg.Wait()
 }
 
-func (s *Scheduler) runTask(t task) {
-	j := t.job
-	s.active++
-
-	go func() {
-		defer func() {
-			if r := recover(); r != nil {
-				logp.Err("Panic in job '%v'. Recovering, but please report this: %s.",
-					j.id, r)
-				logp.Err("Stacktrace: %s", debug.Stack())
-				s.signalFinished(j, nil)
-			}
-		}()
+// runRecursiveTask runs an individual task and its continuations until none are left with as much parallelism as possible.
+// Since task funcs can emit continuations recursively we need a function to execute
+// recursively.
+// The wait group passed into this function expects to already have its count incremented by one.
+func (s *Scheduler) runRecursiveTask(jobCtx context.Context, task TaskFunc, wg *sync.WaitGroup) {
+	defer wg.Done()
 
-		cont := t.fn()
-		s.signalFinished(j, cont)
-	}()
-}
+	// The accounting for waiting/active tasks is done using atomics.
+	// Absolute accuracy is not critical here so the gap between modifying waitingTasks and activeJobs is acceptable.
+	s.waitingTasks.Inc()
 
-func (s *Scheduler) addSync(j *job) {
-	j.registered = true
-	s.jobs = append(s.jobs, j)
-}
+	// Acquire an execution slot in keeping with heartbeat.scheduler.limit
+	s.sem.Acquire(s.ctx, 1)
+	defer s.sem.Release(1)
 
-func (s *Scheduler) doRemove(j *job) {
-	// find entry
-	idx := -1
-	for i, other := range s.jobs {
-		if j == other {
-			idx = i
-			break
-		}
-	}
-	if idx == -1 {
+	// Check if the scheduler has been shut down. If so, exit early
+	select {
+	case <-jobCtx.Done():
 		return
-	}
+	default:
+		s.activeTasks.Inc()
+		s.waitingTasks.Dec()
 
-	// delete entry, not preserving order
-	s.jobs[idx] = s.jobs[len(s.jobs)-1]
-	s.jobs = s.jobs[:len(s.jobs)-1]
+		continuations := task()
+		s.activeTasks.Dec()
 
-	// mark entry as unregistered
-	j.registered = false
-}
-
-func (s *Scheduler) signalFinished(j *job, cont []TaskFunc) {
-	var tasks []task
-	if len(cont) > 0 {
-		tasks = make([]task, len(cont))
-		for i, f := range cont {
-			tasks[i] = task{j, f}
+		for _, cont := range continuations {
+			go s.runRecursiveTask(jobCtx, cont, wg) // Run continuations in parallel, note that these each will acquire their own slots
```

Comment:
Looks like we are missing calls to `wg.Add`. If the children go-routines call  `wg.Done` then the counter becomes negative and you have a panic.

No these recursive calls and proper usage of `sync.WaitGroup` become tricky. One must not call `Add` of some other go-routine is blocked on `Wait`. Otherwise the `Add` will panic. In order to resolve this one needs a local `sync.WaitGroup` and wait on it until all children are done. This way the 'finish' of a computation creates a cascade of go-routines shutting down one after anohter.

A waitGroup kind of acts like a barrier, but one must not add children while waiting for it.

Your use-case (quit finish go-routines before child go-routines) can be implemented using a reference counter + channel. Using the reference counter from [github.com/elastic/go-concert](https://github.com/elastic/go-concert/blob/master/refcount.go) the pattern becomes:

```
waitCh := make(chan struct{})
refCount := &concert.RefCount{
  Action: func(_ error) { close(waitCh) },
}

refCount.Retain() // add ownership for go-routine
go s.runRecursiveTask(jobCtx, task, refCount)

recCount.Release() // give up local ownership
<-waitCh // wait for all child go-routines to finish
```

with:

```
func (s *Scheduler) runRecursiveTask(jobCtx context.Context, task TaskFunc, refCount *concert.RefCount) {
  defer refCount.Release()
  
  ...
  for _, cont := range continuations {
    refCount.Retain() / add ownership for new go-routine
    go s.runRecursiveTask(jobCtx, cont, refCount)
  }
}
```

Reply:
You're right! I'm missing an `Add` for the WaitGroup.

Are you sure that's true about `WaitGroup` though? In the [docs](https://golang.org/pkg/sync/#WaitGroup.Add) it says:

>  Note that calls with a positive delta that occur when the counter is zero must happen before a Wait. Calls with a negative delta, or calls with a positive delta that start when the counter is greater than zero, may happen at any time. 

The ref count approach is neat, but I think it's more complicated, and I'd rather not use it unless I have to.

Reply:
I've added the missing `add` and it seems to work fine. I'm realizing I need to add scheduler tests with continuations however.

Reply:
+1 on testing. Right, one must not run the first call to `Add` concurrently with `Wait`, but if the count counter is > 0 when calling `Add` it should be safe. When starting child go-routines waitGroup kind of should be used as a reference counter: Always call Add before giving up on the waitgroup by calling done.

Reply:
100% right. I think we're OK here since we increment the add counter before invoking the continuations.